File: 2487.从链表中移除节点.cpp
/*
 * @lc app=leetcode.cn id=2487 lang=cpp
 *
 * [2487] 从链表中移除节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *reverseNodes(ListNode *head)
    {
        ListNode *prev = nullptr;
        while (head != nullptr)
        {
            ListNode *nxt = head->next;
            head->next = prev;
            prev = head;
            head = nxt;
        }
        return prev;
    }
    ListNode *removeNodes(ListNode *head)
    {
        /**
         * 5 2 13 3 8 移除右侧有更大结点
         * 5 -> 13 和 8
         * 2 -> 13 3 和 8
         * 13 保存
         * 3 -> 8
         * 8  保存
         * 最后一个结点一定被保存
         * 13 和 8
         * 如果我们反转链表
         * 8 -> 3 -> 13 -> 2 -> 5
         * 我们如何来计算呢?
         * 只需要保存 比左边大的值即可?
         */
        // 反转了链表
        // 然后双指针进行比较吗?
        head = reverseNodes(head);
        ListNode *cur = head;
        while (cur->next)
        {
            if (cur->val > cur->next->val)
            {
                cur->next = cur->next->next; // 链接上了
            }
            else
            {
                cur = cur->next;
            }
        }
        return reverseNodes(head);
    }
};
// @lc code=end

File: 2487.从链表中移除节点.py
#
# @lc app=leetcode.cn id=2487 lang=python3
#
# [2487] 从链表中移除节点
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        5 2 13 3 8
        5 13 8
        2 13 3 8
        13 保存
        3 8
        从右往左递归，node 始终代表“右侧子链表的最大值所在的节点”；
        若这个最大值大于当前值，删当前节点，否则保留并接回
        """
        # 递归基
        if head.next == None:
            return head
        node = self.removeNodes(head.next)
        if node.val > head.val:
            # 如果右侧子链表的最大值比当前值大,说明当前节点应该被删除
            return node
        head.next = node
        return head
        
# @lc code=end

